### ISSUE TITLE
Cannot use ASP.NET Core without using Entity Framework.

### ISSUE THREAD

**Issue by eabemitchell** [dotnet/aspnetcore#35381]
**My Person reasons:**
I understand others love the EF (Entity Framework). Usually new programmers who don't understand databases. However; I find it much more limiting and not very practical in most development environments. I was trying to use SQLClient with ASP.Net Core 5 and it just sucks that all the asp and session functionality was basically removed. And instead a huge focus forcing you to use EF. And forcing me to use it to validate data on pages is even more troubling. 

**Brief examples as to why:**
I do not want to say a database field is always required. On this page in this scenario it is. On this page in this scenario it is not. If the person is a student vs a teacher I want the required fields to be different. Such as their SSN... For payroll purposes. Students are not paid by the school so is not required for taxes. I really don't want to have to create separate unneeded tables for people. EF is incredibly limiting in many ways. And the fact I can't get data to carry over to the controller from the cshtml page without the variable names matching exactly is also incredibly frustrating.

I worked in the support depart of my companying for many years. Before recently quitting. And I inherited many bad projects that I was asked to save. Often being put on dying projects with varying languages. And not once was I ever able to save a project that used EF. And EF projects were always the worse too look at and to try and detangle. I had to rewrite the entire code base every single time. Believe me I tried to save them.

I was even specifically asked not to rewrite it a couple of times. Mainly because of time constraints. I spent a month on a project that I was told was suppose to be an easy fix. Everyone on the team quite after inheriting that EF garbage. And the original team was fired prior to that for not making the dead line. My understand it was multiple failed times. And again the project failed as they didn't let me rewrite. Like I said it needed too after reviewing the code for a week. It just couldn't be done the way they wanted. They brought in more people to no avail. It is just to limiting. Eventually that client was lost. EF is only good for very simple and small projects. And to this day is the only reason any project that I worked on didn't pan out.

I really hope ASP.Net core goes back to supporting things like <asp:RequiredFieldValidator> Instead of forcing me to make a database column always required in a EF like class. I hate it, and it is not flexible and forces me to recreate the object just to change it from being required here but not here.

**Conclusion:**
If this is the future of ASP.NET Core you can count me out. I will stick with the older versions. After running into the required field problem. I have decided to scrap all my current code and to re-make my project in ASP.NET and to probable never use core again. It sucks and leaves a very bad taste. And somewhat feels like it dumbs down objects oriented programing in general. To the point that the objects are not really separate objects anymore. Almost like going backwards in time. I call this a regression to ASP.Net Core as it definitely doesn't feel like progression. While it is fine to support EF. To force programmers to use it seems down right stupid. EF can't even support none Microsoft databases that use ODBC? I call ASP.NET Core 5 a failure mostly because of its reliance on EF. What a waste of a week. I was told core was made to support everything. I guess that's what I get for trying to learn new things. I even found a work around to get objects into the session state again using java. What a shame. Why did ASP.NET Core 5 have to be so problematic and future dependent? I thought it was made to run on any OS, and to be flexible. But I find the EF side of it stiffens the code right up.

**Comment by davidfowl** [comment]
It's unclear after reading this several times where you're forced to use EntityFramework.

**Comment by eabemitchell** [comment]
> First off: In our company experienced developers and DBA's work daily with EF Core and are quite satisfied to say at least. This also applies to complex and large databases/projects.
> 
> Regarding the required issue. If I understand correctly you are not able to differ between database required fields and view required fields. Maybe you could try creating a view model class instead of using the database models directly (since there is a validation difference this would also make your code more readable to colleague developers) .
> On the controller/logic side, you can map between the classes using (by example) Automapper or writing a property mapping yourself.
> 
> Another option would be to edit the validation result in the controller using the ModelState...
> 
> Good luck!

Sigh... I never said some people don't like it. I absolutely hate it. That is mostly an opinion thing. But I have my reasons. And ones at that. However; that isn't what is being discussed here.

Anyways your guess isn't really all that correct. Or maybe you are unfamiliar with Asp.Net Core 5. And are using an older version where view and models were much more separated code wise. Here is how Microsoft explains you do it in 5 https://docs.microsoft.com/en-us/aspnet/core/mvc/models/validation?view=aspnetcore-5.0.

As you can see you need to mark the EF object as required and than tell the page to validate it. The old school page validation stuff doesn't seem to work at all in ASP.NET Core 5. Where I passed the view a datatable instead. And it wasn't easy to get the datatable to the cshtml page either.

Whatever; I gave the view an empty datatable with the structure I wanted. I than want to do stuff like mark the field I am creating in the view required. And not do it to the controller based off a ling nested if else. I mean I could do that. However; cshtml built in way of validation in version 5 forces you to do it in the model. Everything I have done in ASP.NET Core 5 has been a work around to get the code to behave the way I want it.

If I could do something like @html.textboxfor(model => DataRow.["column"], thisFieldIsRequired) or something. Than I wouldn't be here complaining.

Whatever; I gave ASP.NET Core 5 a try. I am not going to waste any more time on it. When other languages like just ASP.NET do it better without all the hassle.

> It's unclear after reading this several times where you're forced to use EntityFramework.
> 
> @eabemitchell can you clarify with some code examples?

Ok while it isn't exactly EF. As I can make a class that holds all the DataRows columns as an object. It defiantly is built to be way easier doing it with EF. Can you even find a single ASP.NET Core 5 example that doesn't use EF? I can't.

Why can't I in the .cshtml page just say this field is required? Because it is 100% focused on the fact that it excepts you to use EF, and say it is required there when you create that EF object. It has become a language that excepts EF. Even the way it passes variables around expects it done that way. While it magically passes variables to the controller as long as the names 100% match. And using asp-route-id="PersonID" doesn't work unless I rename PersonID to exactly match the param name of id. Which when you do that rename also makes attribute asp-route-id not needed. As you pretty much loose the ability to name variables differently from .cshtml to the controller. That's also why they removed objects from the session. That's why TempData breaks if you put an object in it. And the list goes on.

If your not using EF with ASP.NET Core 5 your going to have a really rough go at it. As the views are really dumbed down.

**Comment by davidfowl** [comment]
Seems like a rant against Object Oriented Programming/DTOs/objects and and Mvc rather than EntityFramework. Are you trying to use data tables everywhere? It's still not clear to me what you're trying to do, some more specific before and after example would help me visualize the concrete problems you're facing.

That said, there are many EF + ASP.NET Core tutorials and it's the default thing that we document but the coupling between the 2 technologies are very loose.

**Comment by eabemitchell** [comment]
> Seems like a rant against Object Oriented Programming/DTOs/objects and and Mvc rather than EntityFramework. Are you trying to use data tables everywhere? It's still not clear to me what you're trying to do, some more specific before and after example would help me visualize the concrete problems you're facing.
> 
> That said, there are many EF + ASP.NET Core tutorials and it's the default thing that we document but the coupling between the 2 technologies are very loose.

I have nothing against O.O.P. Anyways I found a different way to make it work just a few mins ago. By doing something like this:
<input type="date" required="required" asp-for="@dataRow["DOB"]"/> or @Html.TextBoxFor(m => @dataRow["UserName"], new { @Value = dataRow["UserName"], @required = "required" })

However; I do have something against making every instance of a class property required. Especially when I only want it required sometimes. Like I said if a person is an employee... such as a teacher in this case. Than I want the SSN required as it is needed for payroll. However; if the person is a student, or even a parent. I don't really need to collect that information at all, and it is up to them weather they enter it or not. And having it tied to EF required attribute is not flexible at all.

As I said my problem is EF. I don't want that column required all the freaking time!!! I have no issue with making classes and objects and passing them around. Hell I would probable even be fine with EF if I could dynamically make a column required here but not there. I also take issue with how I cannot natively pass an object to a .cshtml class without a roundabout work around using javascript. And how variables are static from .cshtml to the controller. I actually find that anti O.O.P. But whatever dude. You do you and I'll do me. Like usually I found a work around. And you can consider this issue closed.

In conclusion I think normal ASP.NET is way better than ASP.NET Core. Maybe only a week of messing with it isn't enough time to do it justice. But from what I have seen it isn't worth any more of my time.
